kv,rpc: Introduce dedicated rangefeed connection class.

miretskiy opened the pull request:
Rangefeeds executing against very large tables may experience
OOMs under certain conditions. The underlying reason for this is that
cockroach uses 2MB per gRPC stream buffer size to improve system throughput
(particularly under high latency scenarios).  However, because the rangefeed
against a large table establishes a dedicated stream for each range,
the possiblity of an OOM exist if there are many streams from which events
are not consumed quickly enough.

This PR does two things.

First, it introduces a dedicated RPC connection class for use with rangefeeds.

The rangefeed connnection class configures rangefeed streams to use less memory
per stream than the default connection class.  The initial window size for rangefeed
connection class can be adjusted by updating `COCKROACH_RANGEFEED_RPC_INITIAL_WINDOW_SIZE`
environment variable whose default is 128KB.

For rangefeeds to use this new connection class,
a `kv.rangefeed.use_dedicated_connection_class.enabled` setting must be turned on.

Another change in this PR is that the default RPC window size can be adjusted
via `COCKROACH_RPC_INITIAL_WINDOW_SIZE` environment variable.  The default for this
variable is kept at 2MB.

Changing the values of either of those variables is an advanced operation and should
not be taken lightly since changes to those variables impact all aspects of cockroach
performance characteristics.  Values larger than 2MB will be trimmed to 2MB.
Setting either of those to 64KB or below will turn on "dynamic window" gRPC window sizes.

It should be noted that this change is a partial mitigation, and not a complete fix.
A full fix will require rework of rangefeed behavior, and will be done at a later time.

An alternative to introduction of a dedicated connection class was to simply
lower the default connection window size.  However, such change would impact
all RPCs in a system, and it was deemed too risky at this point.
Substantial benchmarking is needed before such change.

Informs #74219

Release Notes (performance improvement): Allow rangefeed streams to use separate
http connection when `kv.rangefeed.use_dedicated_connection_class.enabled` setting
is turned on.  Using separate connection class reduces the possiblity of OOMs when
running rangefeeds against very large tables.  The connection window size
for rangefeed can be adjusted via `COCKROACH_RANGEFEED_INITIAL_WINDOW_SIZE` environment
variable, whose default is 128KB.

Commit:
changefeedccl: Have null sink record stats.

Record stats while discarding data.

Release Notes: None

Commit:
kv,rpc: Introduce dedicated rangefeed connection class.

Rangefeeds executing against very large tables may experience
OOMs under certain conditions. The underlying reason for this is that
cockroach uses 2MB per gRPC stream buffer size to improve system throughput
(particularly under high latency scenarios).  However, because the rangefeed
against a large table establishes a dedicated stream for each range,
the possiblity of an OOM exist if there are many streams from which events
are not consumed quickly enough.

This PR does two things.

First, it introduces a dedicated RPC connection class for use with rangefeeds.

The rangefeed connnection class configures rangefeed streams to use less memory
per stream than the default connection class.  The initial window size for rangefeed
connection class can be adjusted by updating `COCKROACH_RANGEFEED_RPC_INITIAL_WINDOW_SIZE`
environment variable whose default is 128KB.

For rangefeeds to use this new connection class,
a `kv.rangefeed.use_dedicated_connection_class.enabled` setting must be turned on.

Another change in this PR is that the default RPC window size can be adjusted
via `COCKROACH_RPC_INITIAL_WINDOW_SIZE` environment variable.  The default for this
variable is kept at 2MB.

Changing the values of either of those variables is an advanced operation and should
not be taken lightly since changes to those variables impact all aspects of cockroach
performance characteristics.  Values larger than 2MB will be trimmed to 2MB.
Setting either of those to 64KB or below will turn on "dynamic window" gRPC window sizes.

It should be noted that this change is a partial mitigation, and not a complete fix.
A full fix will require rework of rangefeed behavior, and will be done at a later time.

An alternative to introduction of a dedicated connection class was to simply
lower the default connection window size.  However, such change would impact
all RPCs in a system, and it was deemed too risky at this point.
Substantial benchmarking is needed before such change.

Release Notes (performance improvement): Allow rangefeed streams to use separate
http connection when `kv.rangefeed.use_dedicated_connection_class.enabled` setting
is turned on.  Using separate connection class reduces the possiblity of OOMs when
running rangefeeds against very large tables.  The connection window size
for rangefeed can be adjusted via `COCKROACH_RANGEFEED_INITIAL_WINDOW_SIZE` environment
variable, whose default is 128KB.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74222)
<!-- Reviewable:end -->

erikgrinaker commented:
Nice find, this approach seems reasonable.

However, it leaves me wondering why the initial window size is set to 2 MB, and why we avoid dynamic sizing. I'm assuming it's for predictable performance and memory usage, but has this specific value been picked for a good reason? Are we not vulnerable to the same problem with e.g. DistSQL flows or other traffic, at a large enough scale?

miretskiy commented:
I believe that is indeed possible.  However, I think (with the exception of gossip), other RPCs are not streaming -- which is nice (i.e. the server doesn't keep sending data while we process previous batch); and I also think that admission control and just limits of how many requests e.g. distsql will send out concurrently might be helping here.  Wrt to the 2MB size -- the linked historical issue has a lot of information; namely https://github.com/cockroachdb/cockroach/pull/35161 seems to indicate that higher limits were needed to get good throughput over high latency links; but that PR was put on hold; @petermattis might have more context.

miretskiy commented:
I'm not seeing differences in throughput: running 2k tpcc warehouse workload; changefeeds into "null" sink.
Before:
<img width="703" alt="Screen Shot 2021-12-23 at 3 06 57 PM" src="https://user-images.githubusercontent.com/54641364/147287456-c822c051-5bae-418e-9cf1-0362ff9d4730.png">

After:
<img width="670" alt="Screen Shot 2021-12-23 at 3 06 45 PM" src="https://user-images.githubusercontent.com/54641364/147287444-0216d560-54fb-44ea-8350-70b34c3f154a.png">

Throughput seems basically the same.

ajwerner commented:
> However, I think (with the exception of gossip), other RPCs are not streaming

This is definitely not true. See https://github.com/cockroachdb/cockroach/blob/775c2b0b0302ebf4b0c8c8515e015b793b8585ce/pkg/sql/execinfrapb/api.proto#L163-L173

miretskiy commented:
Ahh. Right. Was looking just one api.proto.  dist flows would likely be O(number of nodes), not num ranges, so there is that

knz commented:
First, a couple general points, given this is opt-in:
What is the strategy here for end-users? For operators? What vision do you have about how this is going to get used?
What criteria should one use to determine whether they should use this and enable it?

Then a couple of technical points:
- Why did you choose 128K? As per your comment on #74219, gRPC dynamic window sizing is available up to 64K. Don't we want dynamic window sizing for rangefeeds? Wouldn't an even smaller window size be desirable, to avoid more stale data from accumulating? Shouldn't the specific size be configurable?
- I'm not too keen on introducing a major new feature wrt networking semantics and _have it disabled by default_. This will prevent it from being exercised by our test suites, nightlies etc.
- I am somewhat surprised that "there is no performance difference". Can you detail the mechanics of your performance test? What is the "null sink"? Are we even sure that the new code was exercised in that benchmark?

miretskiy commented:
Thanks @knz for thoughtful reply.
Need to figure out how to best get this code backported, without adding too much craft 

> First, a couple general points, given this is opt-in: What is the strategy here for end-users? For operators? What vision do you have about how this is going to get used? What criteria should one use to determine whether they should use this and enable it?

I'd rather not have it opt in. However, I'm not too comfortable backporting without kill switch. Basically, operators who experience range feed related ooms could opt in.
> 
> Then a couple of technical points:
> 
> * Why did you choose 128K? As per your comment on [changefeed,kv: Excessive memory usage when using rangefeeds. #74219](https://github.com/cockroachdb/cockroach/issues/74219), gRPC dynamic window sizing is available up to 64K. Don't we want dynamic window sizing for rangefeeds? Wouldn't an even smaller window size be desirable, to avoid more stale data from accumulating? Shouldn't the specific size be configurable?

We probably don't want dynamic size because it could make problem even worse. Namely: the max buffer is 4mb when using dynamic window. However, once we change feeds to use one stream per node, then we should almost certainly use dynamic sizing .  128 was rather randomly chosen. Makes ooms 15x less likely.

> * I'm not too keen on introducing a major new feature wrt networking semantics and _have it disabled by default_. This will prevent it from being exercised by our test suites, nightlies etc.


I hear you. I'm simply not comfortable backporting without kill switch. I did use metamorphic setting so that we would exercise this code in tests .  Would you rather have it on by default, with an opt out? 

> * I am somewhat surprised that "there is no performance difference". Can you detail the mechanics of your performance test? What is the "null sink"? Are we even sure that the new code was exercised in that benchmark?

Well, null sink does every the regular sink does, it just writes to dev/null. In a sense, this is the happiest case. Buffers on a stream exists to improve throughput. As such, lowering this size might effect throughput in a happy case. My very minimal test seems to indicate that that's not the case. The test ran against a 1T table - a reasonably sized workload.  I'm sure the test exercised the right code.

bdarnell reviewed (commented):
> Are we not vulnerable to the same problem with e.g. DistSQL flows or other traffic, at a large enough scale?

Yes, I think so. 

> Wrt to the 2MB size -- the linked historical issue has a lot of information; namely #35161 seems to indicate that higher limits were needed to get good throughput over high latency links

Two additional points: 
- This was using our old `block_writer` benchmark which just does bulk writes of large blobs (it was subsequently replaced by `kv0`); this kind of workload not something we focus on these days . (I believe this even predates geo-partitioning so we were trying to support high-latency links in a particularly naive way)
- At the time that we introduced the 2MB limit, GRPC did not have any kind of dynamic sizing. 

I think a 2MB-per-stream default limit is a stability risk in all cases except perhaps snapshots (which are governed by a higher-level limit on the number of active snapshot streams). And I don't think it's very beneficial for most workloads (even if it does increase throughput for some benchmarks, we should be willing to take small decreases in performance for an improvement in stability). So before introducing a new connection class for rangefeeds, I would consider just making the default connection class's initial window size configurable (which would be backportable) and smaller (in the next release).

There is an alternative approach of primarily relying on connection-level limits instead of stream-level limits, which is how GRPC is configured by default. However, we have run into various deadlock scenarios when streams get blocked by the connection-level limits, so we currently set the connection-level limits "impossibly" high. Connection-level limits are easier to reason about and tune, so it would be nice if we could use them, but resolving those deadlocks could be difficult. 

Our lack of connection-level limits is also a reason to be wary of dynamic sizing. If those limits were in place, they'd keep us from having two many streams with windows expanded to the maximum. But without them, we could have trouble (or maybe not? I don't know exactly how the dynamic sizing works and how the various factors interact). 

> Can you detail the mechanics of your performance test?

Specifically, what was the latency between nodes? The window size matters most for high-latency connections. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-Ms0veIkCozmc9dR33pr:bgk3s6u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @erikgrinaker, @knz, @miretskiy, and @nvanbenschoten)

___
*[pkg/rpc/context.go, line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms1BUp3E4wucn-2yoYJ:-Ms1BUp3E4wucn-2yoYK:b-85svuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c6909c7be64713221285f22a8c9cc2e748e3e4/pkg/rpc/context.go#L79)):*
> ```Go
> // rangefeedInitialWindowSize is the initial window size for rangefeed streaming RPCs.
> var rangefeedInitialWindowSize = func() int32 {
> 	s := envutil.EnvOrDefaultInt("COCKROACH_RANGEFEED_RPC_INITIAL_WINDOW_SIZE", 2*defaultWindowSize /* 128K */)
> ```

If a user sets this env var to 64KiB, does that automagically change to dynamic sizing, or would it use a fixed 64K buffer? (Either answer is acceptable to me but we should know and document which way it is)


<!-- Sent from Reviewable.io -->

miretskiy commented:
> > Are we not vulnerable to the same problem with e.g. DistSQL flows or other traffic, at a large enough scale?
> 
> Yes, I think so.
> 
> > Wrt to the 2MB size -- the linked historical issue has a lot of information; namely #35161 seems to indicate that higher limits were needed to get good throughput over high latency links
> 
> Two additional points:
> 
> * This was using our old `block_writer` benchmark which just does bulk writes of large blobs (it was subsequently replaced by `kv0`); this kind of workload not something we focus on these days . (I believe this even predates geo-partitioning so we were trying to support high-latency links in a particularly naive way)
> * At the time that we introduced the 2MB limit, GRPC did not have any kind of dynamic sizing.
> 
> I think a 2MB-per-stream default limit is a stability risk in all cases except perhaps snapshots (which are governed by a higher-level limit on the number of active snapshot streams). And I don't think it's very beneficial for most workloads (even if it does increase throughput for some benchmarks, we should be willing to take small decreases in performance for an improvement in stability). So before introducing a new connection class for rangefeeds, I would consider just making the default connection class's initial window size configurable (which would be backportable) and smaller (in the next release).
> 

Making buffer size configurable is certainly a possibility.  I was apprehensive about taking this approach
given long historical context to these constants ( linked in the issues).  Isolating rangefeeds seemed a bit less dangerous
to me -- though I maybe wrong.  Certainly, just changing an env variable seems to be very simple and "opt in".

On a more "philosophical" level: rangefeeds do seem to be very different from every other method -- they are meant
to run roughly forever.  Ability to configure such connections differently from other short lived requests might not be a bad thing.

> There is an alternative approach of primarily relying on connection-level limits instead of stream-level limits, which is how GRPC is configured by default. However, we have run into various deadlock scenarios when streams get blocked by the connection-level limits, so we currently set the connection-level limits "impossibly" high. Connection-level limits are easier to reason about and tune, so it would be nice if we could use them, but resolving those deadlocks could be difficult.
> 

Yeah -- those historical issues were pretty scary.  I am working on a change to have 1 rangefeed stream per node (instead of per range).  This way, I can implement sort of connection-level limits: for example, reserve buffer_size*num_nodes memory beforehand.  It's a bit pessimistic, but probably fine since number of nodes is much smaller than ranges.

> Our lack of connection-level limits is also a reason to be wary of dynamic sizing. If those limits were in place, they'd keep us from having two many streams with windows expanded to the maximum. But without them, we could have trouble (or maybe not? I don't know exactly how the dynamic sizing works and how the various factors interact).
> 

Not sure about dynamic sizing either. My read was that it seemed to work fine, but it didn't really fix
the underlying (deadloock) issues -- and so it was put on hold.  

> > Can you detail the mechanics of your performance test?
> 
> Specifically, what was the latency between nodes? The window size matters most for high-latency connections.
> 

My benchmark wasn't focused on high latency at all.  Quite the opposite -- the happy case, 0 latency basically.
The thing about this change is that w/ changefeeds, what matters is that the downstream sink consumes data quickly enough -- something we can't guarantee.  And, so, even w/ pushback in changefeed enabled, the rangefeed portion
happily kept ingesting up to 2 MB of data.  
FOr my benchmark, I wanted to make sure I don't impact existing "happy" workloads.  So, a 1 TB table, changefeed started w/ few hours back in time (to trigger large catchup scan), plus an ongoing tpcc workload.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-Ms0veIkCozmc9dR33pr:bgk3s6u)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @erikgrinaker, @knz, @miretskiy, and @nvanbenschoten)
> 
> _[pkg/rpc/context.go, line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms1BUp3E4wucn-2yoYJ:-Ms1BUp3E4wucn-2yoYK:b-85svuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c6909c7be64713221285f22a8c9cc2e748e3e4/pkg/rpc/context.go#L79)):_
> 
> > ```go
> > // rangefeedInitialWindowSize is the initial window size for rangefeed streaming RPCs.
> > var rangefeedInitialWindowSize = func() int32 {
> > 	s := envutil.EnvOrDefaultInt("COCKROACH_RANGEFEED_RPC_INITIAL_WINDOW_SIZE", 2*defaultWindowSize /* 128K */)
> > ```
> 
> If a user sets this env var to 64KiB, does that automagically change to dynamic sizing, or would it use a fixed 64K buffer? (Either answer is acceptable to me but we should know and document which way it is)

Yup; setting to lless 64K would make it dynamic; I think it's fine and we might actually play with this.

knz reviewed (commented):
One thing I'm not clear on: when there's more than 1 rangefeed between a pair of nodes, this mechanism using connection classes is still going to share 1 gRPC connection (with the new class) across all the rangefeeds for that pair of nodes. Is that desirable?  

> FOr my benchmark, I wanted to make sure I don't impact existing "happy" workloads. 

But the happy workloads are not the problem being solved, right? We want to improve the situation for cases where the sink is jittery, to avoid stale data accumulating in the rangefeed. What's the benchmark for that?

The problem with a benchmark using the null sink is that it does not exercise the effects of backpressure on the workload. I would be interested to compare before/after of a sink that sends the data to some place where the data is written, possibly with some I/O jitter.

> Making buffer size configurable is certainly a possibility. I was apprehensive about taking this approach
given long historical context to these constants ( linked in the issues). Isolating rangefeeds seemed a bit less dangerous
to me -- though I maybe wrong. 

Regardless of the approach taken, a discussion of the choice and the alternatives needs to be included in the commit message. 

> On a more "philosophical" level: rangefeeds do seem to be very different from every other method -- they are meant
to run roughly forever. Ability to configure such connections differently from other short lived requests might not be a bad thing.

This argument holds for incremental backups, exports, and other things that pump data out longer than small requests.
I'd like to remind you / point out that that's what gRPC was supposed to provide us: support for different types of traffic over a single TCP connection. That's the philosophical level: gRPC should be the solution, and we shouldn't need your PR. And yet we are. I think the explanation should step out of philosophy and be more specific about how rangefeeds can't use gRPC adequately.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-Ms5CqI14W8HM7ncr6Lq:bkvjkd1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @erikgrinaker, @miretskiy, and @nvanbenschoten)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms5ESLDElS2odcixblD):*
Right now the commit message is defective in 4 ways:

- 3 ways that are mandatory for every type of work:
  - it must state the problem being solved
  - it must explain how the situation was before your commit
  - it must explain how the solution is helping with the problem
- 1 additional way for work that contains a non-trivial trade-off:
  - it should explain the alternatives you've considered, and why you've chosen this one above others

it would be good to expand this.
___
*[-- commits, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms5D-6S9j6Q834nELhI):*
In addition to the general problems about the commit message, this release note is malformed.
1. The prefix is invalid (should be `Release note (....):`).
2. The mechanism needs to be explained in the release note so that the doc writer has a story to say.
3. The env var needs to be explained too, as well as the mention about the switch to dynamic sizing under 64K.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> One thing I'm not clear on: when there's more than 1 rangefeed between a pair of nodes, this mechanism using connection classes is still going to share 1 gRPC connection (with the new class) across all the rangefeeds for that pair of nodes. Is that desirable?

Correct.  It's not different from today. 

> 
> > FOr my benchmark, I wanted to make sure I don't impact existing "happy" workloads.
> 
> But the happy workloads are not the problem being solved, right? We want to improve the situation for cases where the sink is jittery, to avoid stale data accumulating in the rangefeed. What's the benchmark for that?

No, I think it's not right.  I *know* the code reduces the memory used by http client -- that's the purpose of this PR -- 
this was demostrated in a linked issue -- by writing a custom test and recording heap profiles.  But I also wanted to make sure that such a change would not negatively effect existing workload (that are not OOMing) and that they still perform adequately, even with reduced buffer size.  And that's what the benchmark attempted to show.

> 
> The problem with a benchmark using the null sink is that it does not exercise the effects of backpressure on the workload. I would be interested to compare before/after of a sink that sends the data to some place where the data is written, possibly with some I/O jitter.

Ohh, I exercised that -- null sink has a "delay" parameter that delays emitting each record by customizable time.

> 
> > Making buffer size configurable is certainly a possibility. I was apprehensive about taking this approach
> > given long historical context to these constants ( linked in the issues). Isolating rangefeeds seemed a bit less dangerous
> > to me -- though I maybe wrong.
> 
> Regardless of the approach taken, a discussion of the choice and the alternatives needs to be included in the commit message.
> 
> > On a more "philosophical" level: rangefeeds do seem to be very different from every other method -- they are meant
> > to run roughly forever. Ability to configure such connections differently from other short lived requests might not be a bad thing.
> 
> This argument holds for incremental backups, exports, and other things that pump data out longer than small requests. I'd like to remind you / point out that that's what gRPC was supposed to provide us: support for different types of traffic over a single TCP connection. That's the philosophical level: gRPC should be the solution, and we shouldn't need your PR. And yet we are. I think the explanation should step out of philosophy and be more specific about how rangefeeds can't use gRPC adequately.

Sure, yet here we are.  I wish we can tweak few things in gRPC.
The reality  is that we have existing large scale customers that are experiencing OOMs.  The configurations for gRPC has been in place since before 2019 . 
The question then is:
which change is safer (and backportable)? An option  such as this one -- where at the cost of additional complexity we reduce the probability of OOM, but do not effect any other RPCs (dist flows, snapshots, etc)? Or an option that just allows one to change default class configuration for all rpcs?  

The simpler option is of course to allow changing default class.  But given how long the existing configuration existed
and the fact that such a change would impact every RPC feels like a more dangerous gamble.  Would we be comfortable in
advising our largest customers to run this w/ buffer size set to 128k or even 64k?  I'm comfortable advising for rangefeeds -- and not so much for other RPCs.

> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-Ms5CqI14W8HM7ncr6Lq:bkvjkd1)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @erikgrinaker, @miretskiy, and @nvanbenschoten)
> 
> _[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms5ESLDElS2odcixblD):_ Right now the commit message is defective in 4 ways:
> 
> * 3 ways that are mandatory for every type of work:
>   
>   * it must state the problem being solved
>   * it must explain how the situation was before your commit
>   * it must explain how the solution is helping with the problem
> * 1 additional way for work that contains a non-trivial trade-off:
>   
>   * it should explain the alternatives you've considered, and why you've chosen this one above others
> 
> it would be good to expand this.
> 
> _[-- commits, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-Ms5D-6S9j6Q834nELhI):_ In addition to the general problems about the commit message, this release note is malformed.
> 
> 1. The prefix is invalid (should be `Release note (....):`).
> 2. The mechanism needs to be explained in the release note so that the doc writer has a story to say.
> 3. The env var needs to be explained too, as well as the mention about the switch to dynamic sizing under 64K.

Sounds good -- I'll address those once bigger questions are answered.  It's quite possible that this PR might have to be dropped in favor of a simpler approach.

Again, thanks for your comments.  Would be interesting to hear what @bdarnell  has to say wrt to the 2 options.

bdarnell reviewed (commented):
> Making buffer size configurable is certainly a possibility.  I was apprehensive about taking this approach
> given long historical context to these constants ( linked in the issues).

To me the extensive debate is a reason *for* configurability. We picked a single number but (I at least) have little confidence that it's the right value for all cases. We have evidence that it produces optimal throughput for certain workloads, but we haven't done much testing with other workloads, nor have we evaluated dimensions other than throughput (i.e. stability) at all. 

> > On a more "philosophical" level: rangefeeds do seem to be very different from every other method -- they are meant
> > to run roughly forever. Ability to configure such connections differently from other short lived requests might not be a bad thing.
> 
> This argument holds for incremental backups, exports, and other things that pump data out longer than small requests.

Philosophically, you're right that rangefeeds are unique in ways that might warrant special connection handling. For example, perhaps their combination of access patterns and latency sensitivity means they'd benefit from a different value of `TCP_NODELAY` than our general traffic. But I'm not sure that that applies to memory budgeting - if a 2MB per stream buffer is a problem, it's a problem no matter what kind of stream it is. 

The best argument I can make for a special memory budget for rangefeeds is that they're the only kind of stream that does a bulk data transfer and goes idle - most streams are either small or move their bulk data at the end and immediately close. This may leave unnecessarily-large buffers lying around in ways that wouldn't be as much of a problem with other traffic, so we could set the buffer sizes separately so that other traffic can get the performance benefits of larger buffers when needed. I don't find this argument very persuasive, though. 

> My benchmark wasn't focused on high latency at all. Quite the opposite -- the happy case, 0 latency basically.

It's nice that this shows no impact on the happiest case, but we could kind of guess that - we know that window size makes relatively little difference with low-latency connections. The environments most likely to have a negative side effect from this change would be those that are running smoothly but have high latency - lowering their window size could limit their throughput, leaving them unable to keep up. 

> The simpler option is of course to allow changing default class. But given how long the existing configuration existed
and the fact that such a change would impact every RPC feels like a more dangerous gamble. Would we be comfortable in
advising our largest customers to run this w/ buffer size set to 128k or even 64k? I'm comfortable advising for rangefeeds -- and not so much for other RPCs.

It's true that advising a lower value for all RPCs will require more testing than just doing it for rangefeeds. But I think we need to do it - I wouldn't want to submit a rangefeed-specific fix to the customer and then have them run into OOMs due to other RPCs running into essentially the same problem. 

I think that introducing a separate service class is, at best, a quick fix to stop the known problem scenario while we do more testing of the general case. Maybe it's worth doing for that reason alone, but if so I would still introduce the env var for modifying the default service class at the same time so we can experiment with both. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-Ms6I0M95TcOXb_hnBo1:bev8c5p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @erikgrinaker, @miretskiy, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

miretskiy commented:
@knz Addressed comments re PR description and release notes

miretskiy commented:
@bdarnell Thanks for comments -- replies inline.


> It's nice that this shows no impact on the happiest case, but we could kind of guess that - we know that window size makes relatively little difference with low-latency connections. The environments most likely to have a negative side effect from this change would be those that are running smoothly but have high latency - lowering their window size could limit their throughput, leaving them unable to keep up.

This is not entirely accurate since changefeeds also have their own (very large buffer) -- but that buffer is accounted for
in memory monitors.  This buffer helps in case of sink slow down (or just high latency connection).  W/ my benchmark I just got a bit more confidence that the loss of grpc buffer (which is between fast crdb nodes) does not pose a problem.

> 
> > The simpler option is of course to allow changing default class. But given how long the existing configuration existed
> > and the fact that such a change would impact every RPC feels like a more dangerous gamble. Would we be comfortable in
> > advising our largest customers to run this w/ buffer size set to 128k or even 64k? I'm comfortable advising for rangefeeds -- and not so much for other RPCs.
> 
> It's true that advising a lower value for all RPCs will require more testing than just doing it for rangefeeds. But I think we need to do it - I wouldn't want to submit a rangefeed-specific fix to the customer and then have them run into OOMs due to other RPCs running into essentially the same problem.
> 
> I think that introducing a separate service class is, at best, a quick fix to stop the known problem scenario while we do more testing of the general case. Maybe it's worth doing for that reason alone, but if so I would still introduce the env var for modifying the default service class at the same time so we can experiment with both.

I agree that it's a quick fix.  I added a note in the PR description to that effect.  Also, introduced ability to change default
window size.

erikgrinaker commented:
> dist flows would likely be O(number of nodes), not num ranges, so there is that

I think it would be O(concurrent queries * number of nodes), because we set up DistSQL streams on a per-query basis. It also has similar characteristics, in that it will schedule e.g. a `Scan` operator on a remote node which will read rows from KV and pass them through the stream as fast as it can. Although there are probably nuances here that the SQL team can better speak to.

>This is not entirely accurate since changefeeds also have their own (very large buffer) -- but that buffer is accounted for
in memory monitors. This buffer helps in case of sink slow down (or just high latency connection).

This seems like what we should do for all consumers that are throughput-sensitive, which would then allow us to reduce all of the stream buffers and have better control over memory usage. I wouldn't be surprised if e.g. DistSQL already did this. We're seeing similar problems with Raft too, that need an equivalent fix.

In the meanwhile, given that some consumers use their own buffers and others don't, perhaps adaptive window sizing would be more suitable.

> It's true that advising a lower value for all RPCs will require more testing than just doing it for rangefeeds. But I think we need to do it - I wouldn't want to submit a rangefeed-specific fix to the customer and then have them run into OOMs due to other RPCs running into essentially the same problem.

I tend to agree with Ben here.

>I agree that it's a quick fix. I added a note in the PR description to that effect. Also, introduced ability to change default
window size.

Thanks. I'm not opposed to this quick fix, but it is fairly heavy-handed and a bit of a slippery slope -- are we just going to do the same thing when customers report OOMs with DistSQL or Raft (which we already see)? Do we have a plan for getting rid of this connection class?

How much would this customer need to reduce the stream buffers by to avoid OOMs?

knz commented:
>  it is fairly heavy-handed and a bit of a slippery slope

Agreed, and also raises a separate system concern: that of the excess TCP connections. With N-1 peer nodes and M connection classes, we need to allocate M * (N-1) TCP ports for the connections. With large clusters, the impact of adding a connection class is non-trivial on the TCP stack in the OS (and possible interference with other services). 

I think if we're finding that gRPC is inadequate in some dimensions it would be worth investigating how to improve gRPC instead of fishing for heavy-handed solutions that will pop up as a resource cost in other places.

erikgrinaker commented:
Random thought: perhaps calling this new connection class `SMALL_BUFFER` or something similar, and allowing it to be shared by consumers that are expected to do their own buffering, would be slightly more palatable and allow us to gradually migrate separate components over to it as we test and tweak them? The end goal would be for this to become the new default class.

erikgrinaker commented:
>Agreed, and also raises a separate system concern: that of the excess TCP connections. With N-1 peer nodes and M connection classes, we need to allocate M * (N-1) TCP ports for the connections. With large clusters, the impact of adding a connection class is non-trivial on the TCP stack in the OS (and possible interference with other services).

I'm not overly worried about this. Modern TCP stacks are extremely efficient, and e.g. web servers routinely service millions of concurrent TCP connections.

miretskiy commented:
> I think it would be O(concurrent queries * number of nodes), 

Sure, I meant per single stream/query/rangefeed.

> This seems like what we should do for all consumers that are throughput-sensitive, which would then allow us to reduce all of the stream buffers and have better control over memory usage. I wouldn't be surprised if e.g. DistSQL already did this. We're seeing similar problems with Raft too, that need an equivalent fix.

ack.

> Thanks. I'm not opposed to this quick fix, but it is fairly heavy-handed and a bit of a slippery slope -- are we just going to do the same thing when customers report OOMs with DistSQL or Raft (which we already see)? Do we have a plan for getting rid of this connection class?
> 
> I think if we're finding that gRPC is inadequate in some dimensions it would be worth investigating how to improve gRPC instead of fishing for heavy-handed solutions that will pop up as a resource cost in other places.

@ericharmeling @knz Both of you bring up the heavy handed approach of this proposed fix.  Wrt to investigating gRPC or trying to improve gRPC -- as discussed before, this is definitely more heavy weight, albeit probably the "right thing to do".

Let's be clear: I am working on a "real" solution (as far as rangefeeds are concerned) -- but this real solution would not apply to other parts of the system -- such as raft.  So, a lot more work (and benchmarking) is needed to make those happen.  I am convinced that this rangefeed solution cannot possibly be backported.

We have an existing customer that experienced OOMs on 21.2.x.  The way I see it, we have few options:
  1. Do nothing.  This is probably not very palatable to that customer since they are trying to use changefeeds at mission critical workloads -- so, OOMing, and possibly loop-ooming the cluster is not ideal.
  2. Make it possible to adjust default window size.  This was discussed above -- are we comfortable changing window size for *all* RPCs?  I'm not.
  3. Something along the lines of this PR -- it's deficiencies, and heavy handedness aside.

So, I think this PR does few things: 1. it makes default window size adjustable as suggested by @bdarnell .  This should enable us to experiment, re-run benchmarks for e.g. raft, etc to come up with a better/alternative configuration.  This will not by itself solve the possibility of OOMs.  More investigation needed.  Perhaps there is a way to hook into rpc stream to detect when it decides to adjust window sizes, and somehow hook that to mem monitor. 
Another thing this PR does (in a heavy handed way) is to mitigate the problem for the existing customer(s) using changefeeds at scale.  If we conduct additional studies, come up wi/ better configs, ects, this new class can easily be removed at a later point.

I understand that @erikgrinaker  is not against merging this.  @knz sounds like you have more (valid, and sensible) concerns.  But given the above options, do you think #1 is preferable? Or perhaps you have other ideas?

> Random thought: perhaps calling this new connection class SMALL_BUFFER or something similar, and allowing it to be shared by consumers that are expected to do their own buffering, would be slightly more palatable and allow us to gradually migrate separate components over to it as we test and tweak them? The end goal would be for this to become the new default class.

Well, on the one hand -- yes, that makes sense.  But given the fact that you can set this env to be very large, even 4MB... I'd say that maybe it's not small buffer? Perhaps soemthing like "STREAMING" -- like the intent is to use this buffer size for streaming rpcs?  Wdyt, @erikgrinaker ?

bdarnell reviewed (approved):
> How much would this customer need to reduce the stream buffers by to avoid OOMs? 

Yes, this is an important question. We may not be able to know for sure, but we have to have the first guess that we're going to test. The less of a change we need to make, the more comfortable we could be applying it to all connection classes. Conversely, do we need to go smaller than 128KB, and if so will dynamic resizing undo this change?

> Are they running in a high-latency environment?

If this is the customer I think it is, then no. 

> Random thought: perhaps calling this new connection class SMALL_BUFFER or something similar, and allowing it to be shared by consumers that are expected to do their own buffering

I think this is premature generalization. We don't know whether there are any other connections that would specifically benefit from small buffers, nor do we know that shrinking the buffer size is the only customization we want to make for rangefeeds. 

> Modern TCP stacks are extremely efficient, and e.g. web servers routinely service millions of concurrent TCP connections.

True, but this tends to require some operating system customizations since limits are often lower by default (e.g. file descriptor limits defaulting to tens of thousands). This isn't the biggest thing we have to worry about but it's possible that increasing the number of TCP connections per node pair could push a cluster over the edge. 


Anyway, I approve this change now that it gives the customer both the option to change the default connection class and to move changefeeds into a separate class. I would suggest that we start with this customer by reducing the window size for the default class and only enable the second class if we discover regressions due to that. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74222#-:-MsaZzx18TD6n7ihhYHU:bj0wc57)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @bdarnell, @erikgrinaker, @gh-casper, @miretskiy, @nvanbenschoten, and @samiskin)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
>I understand that @erikgrinaker is not against merging this.

Nah, I think this is fine, if a bit unfortunate.

>Wrt to investigating gRPC or trying to improve gRPC -- as discussed before, this is definitely more heavy weight, albeit probably the "right thing to do".

nit: maybe this is what you and @knz meant, but it's specifically gRPC-Go that's the problem, not gRPC itself. HTTP/2 supports per-stream window sizes just fine: https://datatracker.ietf.org/doc/html/rfc7540#section-5.2. But yeah, we have to either fork/extend gRPC-Go or use separate connections if we want to vary stream parameters.

>Make it possible to adjust default window size. This was discussed above -- are we comfortable changing window size for all RPCs? I'm not.

This is going to be workload- and environment-specific though, and I don't think anything would completely _break_ -- it would "just" result in poorer performance in the worst case. So I don't think adding a knob and asking them to try to tweak it for their environment is a terrible option. If they're on a low-latency network it might well resolve the problem for them without any significant drawbacks.

>Well, on the one hand -- yes, that makes sense. But given the fact that you can set this env to be very large, even 4MB... I'd say that maybe it's not small buffer? Perhaps soemthing like "STREAMING" -- like the intent is to use this buffer size for streaming rpcs? Wdyt, @erikgrinaker ?

Yeah, that'd work. We'd just want to group together streams that need the same characteristics into each class. This assumes that all streaming RPCs all want the same characteristics/parameters -- is that always true?

>I think this is premature generalization. We don't know whether there are any other connections that would specifically benefit from small buffers, nor do we know that shrinking the buffer size is the only customization we want to make for rangefeeds.

Good point. I also think we can rename these classes (or move them) later without significant backwards compatibility concerns, so we probably don't need to settle this once and for all right now. A rangefeed class seems fine for now.

>Anyway, I approve this change now that it gives the customer both the option to change the default connection class and to move changefeeds into a separate class. I would suggest that we start with this customer by reducing the window size for the default class and only enable the second class if we discover regressions due to that.

+1

knz commented:
> it's specifically gRPC-Go that's the problem, not gRPC itself. HTTP/2 supports per-stream window sizes just fine: datatracker.ietf.org/doc/html/rfc7540#section-5.2. But yeah, we have to either fork/extend gRPC-Go or use separate connections if we want to vary stream parameters.

@miretskiy can you file a followup issue to do just this, and assign it to the server team project, and cc me? Thanks
